ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ADAuthBasedStorageClient.java

```diff
@@ -135,14 +158,26 @@ protected boolean handleExceptionAndHintRetry(BlobStorageException blobStorageEx
         // to attempt token refresh at the same time. It is expected that as a result of token refresh, accessTokenRef
         // will updated with the new token.
         if (accessTokenRef.get().isExpired()) {
-          azureMetrics.absTokenRefreshAttemptCount.inc();
-          BlobServiceClient blobServiceClient = createBlobStorageClient();
-          setClientReferences(blobServiceClient);
-          logger.info("Token refresh done.");
+          refreshStorageClient();
         }
       }
       return true;
     }
     return false;
   }
+
+  /**
+   * Refreshes AD authentication token and creates a new ABS client with the new token.
+   */
+  private synchronized void refreshStorageClient() {
+    // if a task is scheduled to refresh token then remove it.
+    if (scheduledFutureRef.get() != null) {
+      scheduledFutureRef.get().cancel(false);
```

Comment:
Is it necessary to cancel and reschedule the refresh task every time?  buildBlobServiceClient can create it only if necessary.

Reply:
I have added the cancel task just to avoid any case where a token might get expired earlier than expected. In that case we don't want token refreshes to happen in quick succession.